src/actions/tabs.js

```diff
@@ -50,11 +60,13 @@ export function moveTab(url: string, tabIndex: number): Action {
  * @memberof actions/tabs
  * @static
  */
-export function closeTab(url: string) {
+export function closeTab(source: Source) {
   return ({ dispatch, getState, client }: ThunkArgs) => {
-    removeDocument(url);
+    const { id, url } = source;
 
-    const tabs = removeSourceFromTabList(getSourceTabs(getState()), url);
+    removeDocument(id);
```

Comment:
is it okay to switch from ID to URL? Would we need to change setDocument to also use ID?

Reply:
also, if we're switching from URL, maybe we want to actually do something smarter... 

```diff

-let sourceDocs: SourceDocuments = {};
+let sourceDocs: SourceDocuments = new WeakMap();

-export function getDocument(key: string) {
-  return sourceDocs[key];
+export function getDocument(source: Source) {
+  return sourceDocs.get(source);
 }

-export function hasDocument(key: string): boolean {
-  return !!getDocument(key);
+export function hasDocument(source: Source): boolean {
+  return sourceDocs.has(source, doc)
 }

-export function setDocument(key: string, doc: any) {
-  sourceDocs[key] = doc;
+export function setDocument(source: Source, doc: any) {
+  sourceDocs.set(source, doc)
 }

-export function removeDocument(key: string) {
-  delete sourceDocs[key];
-}
-
-export function clearDocuments() {
-  sourceDocs = {};
+export function removeDocument(source: Source) {
+  sourceDocs.remove(source);
 }
```

this means we won't ever need to call `clearDocuments`

Reply:
I like this idea but I think we should do this as a follow up; this PR is big enough already and `WeakMap` doesn't have a solution for clearing itself, so I don't think your `this means we won't ever need to call clearDocuments` statement is true.

Reply:
This is the only case I saw `removeDocument` used with a `url`; every other function uses a source ID